src/services/helpers/workspaceAdapter.ts

```diff
@@ -0,0 +1,229 @@
+  IDevWorkspaceDevfile,
+} from '@eclipse-che/devworkspace-client';
+import { attributesToType, typeToAttributes } from '../storageTypes';
+import { DevWorkspaceStatus, WorkspaceStatus } from './types';
+
+export interface Workspace {
+  readonly ref: che.Workspace | IDevWorkspace;
+
+  readonly id: string;
+  name: string;
+  readonly namespace: string;
+  readonly infrastructureNamespace: string;
+  readonly created: number;
+  readonly updated: number;
+  status: keyof typeof WorkspaceStatus | keyof typeof DevWorkspaceStatus;
+  readonly ideUrl?: string;
+  devfile: che.WorkspaceDevfile | IDevWorkspaceDevfile;
+  storageType: che.WorkspaceStorageType;
+  readonly projects: string[];
+}
+
+export class WorkspaceAdapter<T extends che.Workspace | IDevWorkspace> implements Workspace {
+  private workspace: T;
+
+  constructor(workspace: T) {
+    if (isWorkspaceV1(workspace) || isWorkspaceV2(workspace)) {
+      this.workspace = workspace;
+    } else {
+      console.error('Unexpected workspace object shape:', workspace);
+      throw new Error('Unexpected workspace object shape.');
+    }
+  }
+
+  get ref(): T {
+    return this.workspace;
+  }
+
+  get id(): string {
+    if (isWorkspaceV1(this.workspace)) {
+      return this.workspace.id;
+    } else {
+      return (this.workspace as IDevWorkspace).status.workspaceId;
+    }
+  }
+
+  get name(): string {
+    if (isWorkspaceV1(this.workspace)) {
+      return this.workspace.devfile.metadata.name || '';
+    } else {
+      return (this.workspace as IDevWorkspace).metadata.name;
+    }
+  }
+
+  set name(name: string) {
+    if (isWorkspaceV1(this.workspace)) {
+      this.workspace.devfile.metadata.name = name;
+    } else {
+      throw new Error('Not implemented.');
+    }
+  }
+
+  get namespace(): string {
+    if (isWorkspaceV1(this.workspace)) {
+      return this.workspace.namespace || '';
+    } else {
+      return (this.workspace as IDevWorkspace).metadata.namespace;
+    }
+  }
+
+  get infrastructureNamespace(): string {
+    if (isWorkspaceV1(this.workspace)) {
+      return this.workspace.attributes?.infrastructureNamespace || '';
+    } else {
+      return (this.workspace as IDevWorkspace).metadata.namespace;
+    }
+  }
+
+  get created(): number {
+    if (isWorkspaceV1(this.workspace)) {
+      return parseInt(this.workspace.attributes?.created || '', 10) || 0;
+    } else {
+      const reference = this.workspace as IDevWorkspace;
+      if (!reference.metadata.creationTimestamp) {
+        return 0;
+      }
+      return new Date(reference.metadata.creationTimestamp).valueOf();
+    }
+  }
+
+  get updated(): number {
+    if (isWorkspaceV1(this.workspace)) {
+      return parseInt(this.workspace.attributes?.updated || '', 10) || 0;
+    } else {
+      const reference = this.workspace as IDevWorkspace;
+      if (!reference.metadata.creationTimestamp) {
+        return 0;
+      }
+      return new Date(reference.metadata.creationTimestamp).valueOf();
```

Comment:
Well, creation is not the same as updated. Probably Dashboard needs to own its own annotation to store updated value. But it's a different story, but we need somewhere to list what is missed for DevWorkspaces